swarm/src/either.rs

```diff
@@ -0,0 +1,248 @@
+            }
+            (Either::Right(behaviour), IntoEitherHandler::Right(handler)) => {
+                behaviour.inject_dial_failure(peer_id, handler, error)
+            }
+            _ => unreachable!(),
+        }
+    }
+
+    fn inject_listen_failure(
+        &mut self,
+        local_addr: &Multiaddr,
+        send_back_addr: &Multiaddr,
+        handler: Self::ProtocolsHandler,
+    ) {
+        match (self, handler) {
+            (Either::Left(behaviour), IntoEitherHandler::Left(handler)) => {
+                behaviour.inject_listen_failure(local_addr, send_back_addr, handler)
+            }
+            (Either::Right(behaviour), IntoEitherHandler::Right(handler)) => {
+                behaviour.inject_listen_failure(local_addr, send_back_addr, handler)
+            }
+            _ => unreachable!(),
+        }
+    }
+
+    fn inject_new_listener(&mut self, id: ListenerId) {
+        match self {
+            Either::Left(a) => a.inject_new_listener(id),
+            Either::Right(b) => b.inject_new_listener(id),
+        }
+    }
+
+    fn inject_new_listen_addr(&mut self, id: ListenerId, addr: &Multiaddr) {
+        match self {
+            Either::Left(a) => a.inject_new_listen_addr(id, addr),
+            Either::Right(b) => b.inject_new_listen_addr(id, addr),
+        }
+    }
+
+    fn inject_expired_listen_addr(&mut self, id: ListenerId, addr: &Multiaddr) {
+        match self {
+            Either::Left(a) => a.inject_expired_listen_addr(id, addr),
+            Either::Right(b) => b.inject_expired_listen_addr(id, addr),
+        }
+    }
+
+    fn inject_new_external_addr(&mut self, addr: &Multiaddr) {
+        match self {
+            Either::Left(a) => a.inject_new_external_addr(addr),
+            Either::Right(b) => b.inject_new_external_addr(addr),
+        }
+    }
+
+    fn inject_expired_external_addr(&mut self, addr: &Multiaddr) {
+        match self {
+            Either::Left(a) => a.inject_expired_external_addr(addr),
+            Either::Right(b) => b.inject_expired_external_addr(addr),
+        }
+    }
+
+    fn inject_listener_error(&mut self, id: ListenerId, err: &(dyn std::error::Error + 'static)) {
+        match self {
+            Either::Left(a) => a.inject_listener_error(id, err),
+            Either::Right(b) => b.inject_listener_error(id, err),
+        }
+    }
+
+    fn inject_listener_closed(&mut self, id: ListenerId, reason: Result<(), &std::io::Error>) {
+        match self {
+            Either::Left(a) => a.inject_listener_closed(id, reason),
+            Either::Right(b) => b.inject_listener_closed(id, reason),
+        }
+    }
+
+    fn poll(
+        &mut self,
+        cx: &mut Context<'_>,
+        params: &mut impl PollParameters,
+    ) -> Poll<NetworkBehaviourAction<Self::OutEvent, Self::ProtocolsHandler>> {
+        let event = match self {
+            Either::Left(behaviour) => futures::ready!(behaviour.poll(cx, params))
+                .map_out(|h| Either::Left(h))
```

Comment:
```suggestion
                .map_out(|e| Either::Left(e))
```

Nit pick. Same for the event mappers below.